jenkins/_analysis.py

```diff
@@ -309,6 +340,102 @@ def get_classified_failures(build_data):
+
+    As a baseline we take the duration of __main_multijob. An engineer should
+    not merge a branch in ideal conditions until __main_multijob is concluded.
+
+    If a job fails, we add to the time of __main_multijob the amount of time of
+    the next run of the same job. In the worst case a job fails and it was the
+    last job to run in __main_multijob. This approximates immediately hitting
+    retry on the failed job in that scenario. This is repeated until we get a
+    to a successful run of the job.
+
+    Note that if the most recent job is a failure, then this column will have
+    NaT in this column.
+
+    :param build_data: A DataFrame representing the individual builds of the
+        test jobs.
+    :returns: A DataFrame that has all of the original columns plus an
+        additional column named "duration_until_mergable" that contains the
+        amount of time from the start of __main_multijob until this sub-build
+        has been retried until success, assuming this was the final job in the
+        __main_multijob build.
+    """
+    grouped = build_data.groupby('job')
+    new_columns = None
+    for name, group in grouped:
+        last = None
+        group_rows = []
+        for row in group.sort_values(by='datetime',
+                                     ascending=False).itertuples():
+            new = {
+                "index": row.Index,
+                "time_to_success": row.sub_duration,
+                "duration_until_mergable": row.duration
+            }
+            if row.result != SUCCESS:
+                if last:
+                    new['time_to_success'] += last["time_to_success"]
+                    new['duration_until_mergable'] += (
+                        last["time_to_success"])
+                else:
+                    new['time_to_success'] = numpy.nan
+                    new['duration_until_mergable'] = numpy.nan
+            last = new
+            group_rows.append(new)
+        new_rows = pandas.DataFrame.from_records(
+            group_rows, index=['index'], exclude=['time_to_success'])
+        if new_columns is None:
+            new_columns = new_rows
+        else:
+            new_columns = new_columns.append(new_rows)
+    result = build_data.join(new_columns)
+    return result
+
+
+def get_daily_time_to_merge(build_data):
+    """
+    Construct a DataFrame that returns a per-day approximation of the amount of
+    time between CI starting and all builds going green. This is an attempt at
+    approximating time to merge.
+
+    :param build_data: A DataFrame with the per-subbuild information.
+    :returns: A DataFrame with one row for each day the data occurred during,
+        and has values that are the average approximated time to submit for
+        that day.
+    """
+    build_data_with_durations = get_time_to_success(build_data)
+
+    def max_preserving_NaTs(x):
+        if any(x.isnull()):
+            return pandas.NaT
+        return numpy.max(x)
```

Comment:
+1. I experimented, was surprised, and added this in.

I guess I probably should have left a test around to show that this is needed. Apologies.